tests/schema/tests.py

```diff
@@ -2391,6 +2395,338 @@ def test_text_field_with_db_index(self):
         assertion = self.assertIn if connection.features.supports_index_on_text_field else self.assertNotIn
         assertion('text_field', self.get_indexes(AuthorTextFieldWithIndex._meta.db_table))
 
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_func_index(self):
+        """
+        Tests creating a functional index
+        """
+
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+
+        # Define the index
+        index_name = 'lower_func_idx'
+        index = Index(fields=[Lower('title').desc()], name=index_name)
+
+        # Add the index
+        with connection.schema_editor() as editor:
+            editor.add_index(Book, index)
+            sql = index.create_sql(Book, editor)
+        self.assertIn(index_name, self.get_constraints(Book._meta.db_table))
+
+        sql = str(sql).upper()
+        self.assertIn('LOWER(', sql)
+        self.assertIn('TITLE', sql)
+        self.assertIn(index_name.upper(), sql)
+        self.assertIn('DESC', sql)
+
+        # Drop the index
+        with connection.schema_editor() as editor:
+            sql = index.remove_sql(Book, editor)
+            editor.remove_index(Book, index)
+        sql = str(sql).upper()
+        self.assertIn(index_name.upper(), sql)
+        self.assertNotIn(index_name, self.get_constraints(Book._meta.db_table))
+
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_composite_func_index(self):
+        """
+        Test creating a composite functional index
+        """
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+
+        # Define the index
+        index_name = 'lower_upper_func_idx'
+        index = Index(fields=[Lower('name'), Upper('name')], name=index_name)
+
+        # Add the index
+        with connection.schema_editor() as editor:
+            editor.add_index(Author, index)
+            sql = index.create_sql(Author, editor)
+
+        # Ensure the index is there
+        self.assertIn(index_name, self.get_constraints(Author._meta.db_table))
+
+        # Check that the vital SQL parts are there
+        sql = str(sql).upper()
+        self.assertIn('LOWER(', sql)
+        self.assertIn('NAME', sql)
+        self.assertIn('UPPER(', sql)
+        self.assertIn(index_name.upper(), sql)
+
+        # Check order of index fields
+        self.assertTrue(
+            sql.index('LOWER') <
+            sql.index('UPPER')
+        )
+
+        # Drop the index
+        with connection.schema_editor() as editor:
+            editor.remove_index(Author, index)
+        self.assertNotIn(index_name, self.get_constraints(Author._meta.db_table))
+
+    @skipUnlessDBFeature('supports_expression_indexes')
+    @unittest.skipIf(
```

Comment:
This seems a bit redundant. I believe that the feature flags should be used as much as possible, instead of checking against vendor names.

Reply:
Ah yes this is should be extracted to a feature flag. Maybe something like `supports_expression_indexes_on_columns`.

MySQL errors with the following message when creating indexes like `Index(fields=[F('title')], name=index_name)` which generates the SQL ```CREATE INDEX `f_func_idx` ON `schema_tag` ((`title`))```:

```
django.db.utils.OperationalError: (3762, 'Functional index on a column is not supported. Consider using a regular index instead.')
```

Reply:
I've added the feature flag `supports_expression_indexes_on_columns`.